x/stake/handler.go

```diff
@@ -0,0 +1,318 @@
+//}
+
+//k.setParams(params)
+//return nil
+//}
+
+//_______________________________________________________________________
+
+func NewHandler(k Keeper, ck bank.CoinKeeper) sdk.Handler {
+	return func(ctx sdk.Context, msg sdk.Msg) sdk.Result {
+		// NOTE msg already has validate basic run
+		switch msg := msg.(type) {
+		case MsgDeclareCandidacy:
+			return k.handleMsgDeclareCandidacy(ctx, msg)
+		case MsgEditCandidacy:
+			return k.handleMsgEditCandidacy(ctx, msg)
+		case MsgDelegate:
+			return k.handleMsgDelegate(ctx, msg)
+		case MsgUnbond:
+			return k.handleMsgUnbond(ctx, msg)
+		default:
+			return sdk.ErrTxDecode("invalid message parse in staking module").Result()
+		}
+	}
+}
+
+//_____________________________________________________________________
+
+// XXX should be send in the msg (init in CLI)
+//func getSender() sdk.Address {
+//signers := msg.GetSigners()
+//if len(signers) != 1 {
+//return sdk.ErrUnauthorized("there can only be one signer for staking transaction").Result()
+//}
+//sender := signers[0]
+//}
+
+//_____________________________________________________________________
+
+// These functions assume everything has been authenticated,
+// now we just perform action and save
+
+func (k Keeper) handleMsgDeclareCandidacy(ctx sdk.Context, msg MsgDeclareCandidacy) sdk.Result {
+
+	// check to see if the pubkey or sender has been registered before
+	_, found := k.GetCandidate(ctx, msg.CandidateAddr)
+	if found {
+		return ErrCandidateExistsAddr().Result()
+	}
+	if msg.Bond.Denom != k.GetParams(ctx).BondDenom {
+		return ErrBadBondingDenom().Result()
+	}
+	if ctx.IsCheckTx() {
+		return sdk.Result{
+			GasUsed: GasDeclareCandidacy,
+		}
+	}
+
+	candidate := NewCandidate(msg.CandidateAddr, msg.PubKey, msg.Description)
+	k.setCandidate(ctx, candidate)
+
+	// move coins from the msg.Address account to a (self-bond) delegator account
+	// the candidate account and global shares are updated within here
+	return k.delegateWithCandidate(ctx, msg.CandidateAddr, msg.Bond, candidate).Result()
+}
+
+func (k Keeper) handleMsgEditCandidacy(ctx sdk.Context, msg MsgEditCandidacy) sdk.Result {
+
+	// candidate must already be registered
+	candidate, found := k.GetCandidate(ctx, msg.CandidateAddr)
+	if !found {
+		return ErrBadCandidateAddr().Result()
+	}
+	if ctx.IsCheckTx() {
+		return sdk.Result{
+			GasUsed: GasEditCandidacy,
+		}
+	}
+	if candidate.Status == Unbonded { //candidate has been withdrawn
+		return ErrBondNotNominated().Result()
+	}
+
+	// XXX move to types
+	//check and edit any of the editable terms
+	if msg.Description.Moniker != "" {
+		candidate.Description.Moniker = msg.Description.Moniker
+	}
+	if msg.Description.Identity != "" {
+		candidate.Description.Identity = msg.Description.Identity
+	}
+	if msg.Description.Website != "" {
+		candidate.Description.Website = msg.Description.Website
+	}
+	if msg.Description.Details != "" {
+		candidate.Description.Details = msg.Description.Details
+	}
+
```

Comment:
Do we care about candidates having the ability to clear previously set metadata?

Reply:
opening an issue

Reply:
https://github.com/cosmos/cosmos-sdk/issues/723